ALICE3/TableProducer/onTheFlyRichPid.cxx

```diff
@@ -0,0 +1,621 @@
+        std::string name_title_barrel_total_res = "h2dBarrelAngularResTotal" + particle_names2[i_true] + "VsP";
+        const AxisSpec axisTrackAngularRes{static_cast<int>(nBinsAngularRes), 0.0f, +5.0f, "Track angular resolution - " + particle_names1[i_true] + " (mrad)"};
+        const AxisSpec axisTotalAngularRes{static_cast<int>(nBinsAngularRes), 0.0f, +5.0f, "Total angular resolution - " + particle_names1[i_true] + " (mrad)"};
+        histos.add(name_title_barrel_track_res.c_str(), name_title_barrel_track_res.c_str(), kTH2F, {axisMomentum, axisTrackAngularRes});
+        histos.add(name_title_barrel_total_res.c_str(), name_title_barrel_total_res.c_str(), kTH2F, {axisMomentum, axisTotalAngularRes});
+      }
+      for (int i_true = 0; i_true < 5; i_true++) {
+        for (int i_hyp = 0; i_hyp < 5; i_hyp++) {
+          std::string name_title = "h2dBarrelNsigmaTrue" + particle_names2[i_true] + "Vs" + particle_names2[i_hyp] + "Hypothesis";
+          if (i_true == i_hyp) {
+            const AxisSpec axisNsigmaCorrect{static_cast<int>(nBinsNsigmaCorrectSpecies), -10.0f, +10.0f, "N#sigma - True " + particle_names1[i_true] + " vs " + particle_names1[i_hyp] + " hypothesis"};
+            histos.add(name_title.c_str(), name_title.c_str(), kTH2F, {axisMomentum, axisNsigmaCorrect});
+          } else {
+            const AxisSpec axisNsigmaWrong{static_cast<int>(nBinsNsigmaWrongSpecies), -10.0f, +10.0f, "N#sigma -  True " + particle_names1[i_true] + " vs " + particle_names1[i_hyp] + " hypothesis"};
+            histos.add(name_title.c_str(), name_title.c_str(), kTH2F, {axisMomentum, axisNsigmaWrong});
+          }
+        }
+      }
+    }
+
+    // Update projective parameters
+    updateProjectiveParameters();
+  }
+
+  /// Function to convert a McParticle into a perfect Track
+  /// \param particle the particle to convert (mcParticle)
+  /// \param o2track the address of the resulting TrackParCov
+  template <typename McParticleType>
+  void convertMCParticleToO2Track(McParticleType& particle, o2::track::TrackParCov& o2track)
+  {
+    // FIXME: this is a fundamentally important piece of code.
+    // It could be placed in a utility file instead of here.
+    auto pdgInfo = pdg->GetParticle(particle.pdgCode());
+    int charge = 0;
+    if (pdgInfo != nullptr) {
+      charge = pdgInfo->Charge() / 3;
+    }
+    std::array<float, 5> params;
+    std::array<float, 15> covm = {0.};
+    float s, c, x;
+    o2::math_utils::sincos(particle.phi(), s, c);
+    o2::math_utils::rotateZInv(particle.vx(), particle.vy(), x, params[0], s, c);
+    params[1] = particle.vz();
+    params[2] = 0.; // since alpha = phi
+    auto theta = 2. * std::atan(std::exp(-particle.eta()));
+    params[3] = 1. / std::tan(theta);
+    params[4] = charge / particle.pt();
+
+    // Initialize TrackParCov in-place
+    new (&o2track)(o2::track::TrackParCov)(x, particle.phi(), params, covm);
```

Comment:
If you are creating a new TrackParCov anyway, why not just return it by value? It would be considerably faster.

Reply:
The codebase is full of code that was not reviewed properly, there is always something that can be improved. Current implementation is, essentially, doing what the copy constructor would do anyway. The better implementation would be a function that simply returns TrackParCov by value to benefit from the return value optimization.